### PR TITLE
EE-1017: Runtime gas optimization

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1017

This PR implements a specialized gas counter object with two separate implementations optimized for separate use cases:

- small counter for `gas_limit <= u64::MAX` that operates on u64 values only
- large counter for `gas_limit > u64::MAX && gas_limit <= U512::MAX` which uses an internal buffer of u64 that gets added up and then is flushed to U512 counter.

There's new benchmark "runtime_bench.rs" which creates a mocked `RuntimeContext` and exercises its methods. It can be further modified to also measure implementations of other host functions.

It's largely based upon earlier research work by @Fraser999 in https://casperlabs.atlassian.net/wiki/spaces/EN/pages/457441767/Optimization+of+Runtime+gas+calls

# Benchmarks

TPS benchmark reports slight improvement (1.5%-3.5%)

```
tps/transfer_to_existing_account_multiple_execs/3/false                                                                            
                        time:   [4.2525 ms 4.2818 ms 4.3041 ms]
                        thrpt:  [697.01  elem/s 700.64  elem/s 705.48  elem/s]
                 change:
                        time:   [-2.8462% -1.5034% -0.3997%] (p = 0.02 < 0.05)
                        thrpt:  [+0.4013% +1.5264% +2.9296%]
                        Change within noise threshold.
tps/transfer_to_existing_account_multiple_deploys_per_exec/3/false                                                                            
                        time:   [4.2498 ms 4.2733 ms 4.3050 ms]
                        thrpt:  [696.86  elem/s 702.03  elem/s 705.92  elem/s]
                 change:
                        time:   [-3.8097% -2.4265% -1.0246%] (p = 0.01 < 0.05)
                        thrpt:  [+1.0352% +2.4868% +3.9605%]
                        Performance has improved.
tps/transfer_to_purse_multiple_execs/3/false                                                                            
                        time:   [4.5748 ms 4.5866 ms 4.5984 ms]
                        thrpt:  [652.41  elem/s 654.08  elem/s 655.76  elem/s]
                 change:
                        time:   [-1.9690% -1.1630% -0.3582%] (p = 0.02 < 0.05)
                        thrpt:  [+0.3595% +1.1767% +2.0086%]
                        Change within noise threshold.
tps/transfer_to_purse_multiple_deploys_per_exec/3/false                                                                            
                        time:   [4.4337 ms 4.4491 ms 4.4666 ms]
                        thrpt:  [671.65  elem/s 674.29  elem/s 676.64  elem/s]
                 change:
                        time:   [-3.5125% -3.0503% -2.5890%] (p = 0.00 < 0.05)
                        thrpt:  [+2.6579% +3.1463% +3.6404%]
                        Performance has improved.
tps/transfer_to_existing_account_multiple_execs/3/true                                                                           
                        time:   [24.290 ms 24.496 ms 24.698 ms]
                        thrpt:  [121.47  elem/s 122.47  elem/s 123.51  elem/s]
                 change:
                        time:   [-3.3672% -1.4868% +0.3594%] (p = 0.15 > 0.05)
                        thrpt:  [-0.3581% +1.5093% +3.4845%]
                        No change in performance detected.
tps/transfer_to_existing_account_multiple_deploys_per_exec/3/true                                                                            
                        time:   [9.3393 ms 9.3905 ms 9.4316 ms]
                        thrpt:  [318.08  elem/s 319.47  elem/s 321.22  elem/s]
                 change:
                        time:   [-2.9857% -1.6272% +0.1201%] (p = 0.07 > 0.05)
                        thrpt:  [-0.1199% +1.6541% +3.0776%]
                        No change in performance detected.
Found 3 outliers among 10 measurements (30.00%)
  1 (10.00%) low mild
  1 (10.00%) high mild
  1 (10.00%) high severe
tps/transfer_to_purse_multiple_execs/3/true                                                                           
                        time:   [23.635 ms 23.815 ms 23.963 ms]
                        thrpt:  [125.19  elem/s 125.97  elem/s 126.93  elem/s]
                 change:
                        time:   [-4.9995% -2.8349% -0.9609%] (p = 0.01 < 0.05)
                        thrpt:  [+0.9702% +2.9177% +5.2626%]
                        Change within noise threshold.
tps/transfer_to_purse_multiple_deploys_per_exec/3/true                                                                            
                        time:   [9.5416 ms 9.6415 ms 9.6990 ms]
                        thrpt:  [309.31  elem/s 311.16  elem/s 314.41  elem/s]
                 change:
                        time:   [-3.2157% -1.4927% -0.0271%] (p = 0.12 > 0.05)
                        thrpt:  [+0.0272% +1.5153% +3.3225%]
                        No change in performance detected.
```

`RuntimeContext::charge_gas` change reports 91% improvement:

```
Gnuplot not found, using plotters backend
gas_charge/65535        time:   [1.2666 ns 1.4671 ns 1.6642 ns]                              
                        change: [-92.598% -91.736% -90.923%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 22 outliers among 100 measurements (22.00%)
  22 (22.00%) high severe
```

### CODE REVIEW

File: execution_engine/src/core/runtime/mod.rs [hunk 1]
```diff
@@ -1427,7 +1425,9 @@ where
         // Charge just for the amount that particular entry point cost - using gas cost from the
         // isolated runtime might have a recursive costs whenever system contract calls other system
         // contract.
-        self.gas(mint_runtime.gas_counter() - gas_counter)?;
+        let gas_diff = mint_runtime.gas_counter().used() - gas_counter.used();
+        debug_assert!(gas_diff.value() <= U512::from(u64::MAX));
+        self.gas(gas_diff.value().as_u64())?;
```
Comment (Fraser999) [inline execution_engine/src/core/runtime/mod.rs:1430]: Is this definitely always going to be safe to cast to a `u64`?  Before `self.gas()` was handling `U512` values.

Maybe we should look to remove the `debug_assert` and just do `self.context.charge_gas_large(gas_diff)?;` here?

File: execution_engine/src/core/runtime/mod.rs [hunk 2]
```diff
@@ -1545,7 +1543,9 @@ where
             _ => CLValue::from_t(()).map_err(Self::reverter),
         };
 
-        self.gas(runtime.gas_counter() - gas_counter)?;
+        let gas_diff = runtime.gas_counter().used() - gas_counter.used();
+        debug_assert!(gas_diff.value() <= U512::from(u64::MAX));
+        self.gas(gas_diff.value().as_u64())?;
```
Comment (Fraser999) [inline execution_engine/src/core/runtime/mod.rs:1548]: Same comment as above re. using `self.context.charge_gas_large(gas_diff)?;`

File: execution_engine/src/core/runtime/mod.rs [hunk 3]
```diff
@@ -1761,7 +1759,11 @@ where
         };
 
         // Charge for the gas spent during execution in an isolated runtime.
-        self.gas(runtime.gas_counter() - gas_counter)?;
+        // Assumes that gas spent does not exceed u64::MAX which as all system contract costs are
+        // u32.
+        let gas_diff = runtime.gas_counter().used() - gas_counter.used();
+        debug_assert!(gas_diff.value() <= U512::from(u64::MAX));
+        self.gas(gas_diff.value().as_u64())?;
```
Comment (Fraser999) [inline execution_engine/src/core/runtime/mod.rs:1766]: Same comment as above re. using `self.context.charge_gas_large(gas_diff)?;`

File: execution_engine/src/shared/gas_counter.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    pub fn add(&mut self, additional_gas: u64) -> Result<(), GasLimitError> {
+        match self {
+            GasCounter::Small(counter) => counter.add(additional_gas),
+            GasCounter::Large(counter) => counter.add(additional_gas),
+        }
+    }
+
+    pub fn add_large(&mut self, additional_gas: Gas) -> Result<(), GasLimitError> {
```
Comment (Fraser999) [inline execution_engine/src/shared/gas_counter.rs:40]: I think it'd be worth documenting why we have two separate methods for adding gas to discourage someone in the future just merging the two `add` methods.

**Review by Fraser999 — CHANGES_REQUESTED** [review]
It'd be good to have all the public methods and types documented, but if that's being deliberately left for future work, no problem.

File: execution_engine/src/shared/gas_counter/large.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    pub(crate) fn add_large(&mut self, mut additional_gas: U512) -> Result<(), GasLimitError> {
+        while additional_gas >= U64_MAX_AS_U512 {
+            self.add(u64::MAX)?;
+            additional_gas -= U64_MAX_AS_U512;
+        }
+        // Safe as we were draining `additional_gas` until it's smaller than `u64::MAX`
+        self.add(additional_gas.as_u64())?;
+        Ok(())
```
[the remark this answers is not in the document]
Comment (mpapierski) [reply execution_engine/src/shared/gas_counter/large.rs]: Good point. I missed this somehow, assuming gas charges wouldn't be anywhere close to `U512::MAX`. Fixed

**Review by Fraser999 — APPROVED** [review]

**Review by henrytill — CHANGES_REQUESTED** [review]
Holding until further notice

File: execution_engine/src/core/runtime/mod.rs [hunk 6]
```diff
@@ -1486,7 +1484,8 @@ where
         // Charge just for the amount that particular entry point cost - using gas cost from the
         // isolated runtime might have a recursive costs whenever system contract calls other system
```
Comment (afck) [inline execution_engine/src/core/runtime/mod.rs:1485]:
```suggestion
        // isolated runtime might have recursive costs whenever system contract calls other system
```
(Or "a recursive cost".)

File: execution_engine/src/core/runtime/mod.rs [hunk 7]
```diff
@@ -1822,7 +1818,10 @@ where
         };
 
         // Charge for the gas spent during execution in an isolated runtime.
-        self.gas(runtime.gas_counter() - gas_counter)?;
+        // Assumes that gas spent does not exceed u64::MAX which as all system contract costs are
+        // u32.
```
Comment (afck) [inline execution_engine/src/core/runtime/mod.rs:1822]: Does it assume that? We're calling `charge_gas_large`, after all?
(On the other hand it probably _would_ be safe to assume it's small, and to call `charge_gas`?)

**Comment by xcthulhu** [comment]
I am not sure we are going to need a large gas counter.

...maybe YAGNI?

**Comment by mpapierski** [comment]
We've discussed that with @Fraser999 and @afck and we can use only small counter and assume `gas_limit <= u64::MAX`

**Comment by mpapierski** [comment]
> 
> 
> We've discussed that with @Fraser999 and @afck and we can use only small counter and assume `gas_limit <= u64::MAX`

This work will happen in separate PR

**Comment by mpapierski** [comment]
Follow up work will happen as part of #1818

**Pull request closed without merging**